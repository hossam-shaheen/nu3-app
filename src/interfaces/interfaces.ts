export interface ProductType {
    id: number;
    title: string;
    price: number;
    image: string;
    vendor: string;
    link: string;
}

export interface ProductsType {
    totalItems: number;
    items: ProductType[];
}

export interface ProductsSearchProps {
    onSearch: (searchKeyWord: string) => void;
}

export interface SortProps {
    onSort: (sortKeyWord: string) => void;
}
export interface ImageType {
    id: number;
    src: string;
    alt: string;
}
export interface ProductAttributeProps {
    productImage: ImageType;
    productImages: ImageType[];
    ProductTitle: string;
    productBodyHtml: string;
    productVendor: string;
}


export interface ProductVariantsProps {
    options: OptionsType[];
    variants: VariantType[];
}


export interface ProductInfoType {
    id: number;
    title: string;
    body_html: string;
    vendor: string;
    images: ImageType[];
    image: ImageType;
    options: OptionsType[];
    variants: VariantType[];
    url: string;
}

export interface VariantType {
    id: number,
    title: string,
    price: number,
}

export interface OptionsType {
    id: number,
    values: string[],
    name: string,
}

export interface ErrorType {
    icon?: string,
    message: string
}