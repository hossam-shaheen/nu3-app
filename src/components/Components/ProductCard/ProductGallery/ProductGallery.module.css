.product-gallery {
    display: flex;
    flex-direction: row-reverse;
    align-items: center;
    width: 100%;
}

.product-gallery-main {
    width: calc(100% - 170px);
}

.product-gallery-main img {
    max-width: 100%;
}

.product-gallery-thumbnails {
    display: flex;
    flex-direction: column;
    width: 25%;
    margin-right: 20px;
}

.product-gallery-thumbnails li {
    width: 100%;
    margin-bottom: 20px;
    cursor: pointer;
}

.product-gallery-thumbnails li img {
    max-width: 100%;
}

.active {
    border: 3px solid var(--secondary);
}