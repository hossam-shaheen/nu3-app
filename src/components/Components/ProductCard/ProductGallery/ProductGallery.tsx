import { FunctionComponent, useEffect, useState } from 'react';
import { ProductAttributeProps } from 'src/interfaces/interfaces';
import classes from './ProductGallery.module.css';


export const ProductGallery: FunctionComponent<{ mainImage: ProductAttributeProps["productImage"] | undefined, thumbnails: ProductAttributeProps["productImages"] | undefined }> = ({ mainImage, thumbnails }): JSX.Element => {

    const [currentMainImage, setCurrentMainImage] = useState<{
        id: number,
        src: string;
        alt: string
    } | null>(null);

    useEffect(() => {
        if (mainImage) {
            setCurrentMainImage({
                id: mainImage.id,
                src: mainImage.src,
                alt: mainImage.alt,
            })
        }
    }, [mainImage]);

    const onChangeThumbnails = (id: number, src: string, alt: string): void => {
        setCurrentMainImage({
            id,
            src,
            alt,
        });
    }

    return <div className={classes["product-gallery"]}>
        <div className={classes["product-gallery-main"]}>
            {currentMainImage && <img src={currentMainImage?.src}
                alt={currentMainImage?.alt} />}
        </div>
        <ul className={classes["product-gallery-thumbnails"]}>
            {thumbnails?.map((image: { id: number; src: string, alt: string }) => {
                return (<li key={image.id}
                    className={currentMainImage?.id === image.id ? classes["active"] : ""}
                    onClick={onChangeThumbnails.bind(null, image.id, image.src, image.alt)}>
                    <img src={image.src} alt={image.alt} />
                </li>)
            })}
        </ul>
    </div>
}
export default ProductGallery;