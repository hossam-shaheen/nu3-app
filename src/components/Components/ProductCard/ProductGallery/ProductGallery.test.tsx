import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { ProductAttributeProps } from "src/interfaces/interfaces";
import ProductGallery from "./ProductGallery";

const mockedMainImage: ProductAttributeProps["productImage"] = {
  id: 4725777367121,
  src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303",
  alt: "nu3 Premium Skin & Hair Beauty",
};

const mockedThumbnails: ProductAttributeProps["productImages"] = [
  {
    id: 4725777367121,
    src: "https://cdn.shopify.com/s/files/1/0095/2274/1329/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303",
    alt: "nu3 Premium Skin Beauty",
  },
  {
    id: 4725777367122,
    src: "https://cdn.shopify.com/s/files/1/0095/2274/1328/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303",
    alt: "nu3 Premium Hair Beauty",
  },
];

test("ProductGallery pass mainImage and thumbnails : Renders", () => {
  render(
    <ProductGallery mainImage={mockedMainImage} thumbnails={mockedThumbnails} />
  );

  const productGalleryThumbnails = screen.queryAllByRole("listitem");
  expect(productGalleryThumbnails.length).toBe(2);
  const productMainImage = screen.queryAllByRole("img")[0];
  expect(productMainImage).toBeInTheDocument();
  expect(productMainImage).toHaveAttribute("alt", mockedMainImage.alt);
});

test(`ProductGallery did't pass mainImage and thumbnails : Renders`, () => {
  render(<ProductGallery mainImage={undefined} thumbnails={undefined} />);
  const productsImages = screen.queryAllByRole("img");
  expect(productsImages.length).toBe(0);
});

test(`ProductGallery click in thumbnails should main image changing : Renders`, () => {
  render(
    <ProductGallery mainImage={mockedMainImage} thumbnails={mockedThumbnails} />
  );

  const productGalleryThumbnails = screen.queryAllByRole("listitem");
  expect(productGalleryThumbnails.length).toBe(2);

  const productMainImage = screen.queryAllByRole("img")[0];

  const firstThumbnail = productGalleryThumbnails[0];
  userEvent.click(firstThumbnail);
  expect(firstThumbnail).toHaveClass("active");
  expect(productMainImage).toHaveAttribute("alt", mockedThumbnails[0].alt);

  const secondThumbnail = productGalleryThumbnails[1];
  userEvent.click(secondThumbnail);
  expect(secondThumbnail).toHaveClass("active");
  expect(productMainImage).toHaveAttribute("alt", mockedThumbnails[1].alt);
});

test("Products pass product : Snapshot", () => {
  const component = renderer.create(
    <ProductGallery mainImage={mockedMainImage} thumbnails={mockedThumbnails} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test(`ProductGallery did't pass mainImage and thumbnails: Snapshot`, () => {
  const component = renderer.create(
    <ProductGallery mainImage={undefined} thumbnails={undefined} />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
