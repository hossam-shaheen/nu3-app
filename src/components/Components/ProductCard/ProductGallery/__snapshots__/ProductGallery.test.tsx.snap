// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductGallery did't pass mainImage and thumbnails: Snapshot 1`] = `
<div
  className="product-gallery"
>
  <div
    className="product-gallery-main"
  />
  <ul
    className="product-gallery-thumbnails"
  />
</div>
`;

exports[`Products pass product : Snapshot 1`] = `
<div
  className="product-gallery"
>
  <div
    className="product-gallery-main"
  />
  <ul
    className="product-gallery-thumbnails"
  >
    <li
      className=""
      onClick={[Function]}
    >
      <img
        alt="nu3 Premium Skin Beauty"
        src="https://cdn.shopify.com/s/files/1/0095/2274/1329/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303"
      />
    </li>
    <li
      className=""
      onClick={[Function]}
    >
      <img
        alt="nu3 Premium Hair Beauty"
        src="https://cdn.shopify.com/s/files/1/0095/2274/1328/products/8_252F1_252Fc_252Fc_252F81ccf47420c1f3a11717ff5ad7f93e3e13214556_21446_large.jpg?v=1620826303"
      />
    </li>
  </ul>
</div>
`;
